### ISSUE TITLE
Docker Compose v2

### ISSUE THREAD

**Issue by chriscroome** [mailcow/mailcow-dockerized#4315]
[Docker Compose v2](https://github.com/docker/compose#docker-compose-v2) is available and [compose-switch](https://github.com/docker/compose-switch) can be used via alternatives on Debian based distros to convert `docker-compose` commands to `docker compose` commands, I realise that the Mailcow `update.sh` script checks the version to use via [this URL](https://www.servercow.de/docker-compose/latest.php) and I don't think this is an urgent or critical issue but I thought it probably deserved a feature request opening as some work might be needed in the future to update the way Mailcow installs and updates Docker Compose.

**Comment by MAGICCC** [comment]
Indeed its released and there's still an issue on compose side with IPv6: https://github.com/docker/compose/issues/8721
Also I will link some closed issues here too: https://github.com/mailcow/mailcow-dockerized/issues/4285 https://github.com/mailcow/mailcow-dockerized/issues/4280

**Comment by nightah** [comment]
https://github.com/docker/compose/issues/8721 can be worked around but surrounding ipv6 literals with square brackets and double quotes, which is common for Go applications.

The key issue which prevented IPv6 routing from working was https://github.com/docker/compose/pull/8851 which has been resolved and merged in a newer version.

**Comment by HorayNarea** [comment]
Now that there are no (known :sweat_smile:) bugs preventing the migration to v2… is there a plan to do this, maybe even a rough estimate when this will happen?

**Comment by StarkZarn** [comment]
It doesn't look like there are any issues preventing this migration. I'll echo what the above poster asked, is there a plan for this? It would increase support for arm-based setups.

**Comment by andryyy** [comment]
Am I missing something? Supporting v2 does help support ARM? ☺️

**Comment by StarkZarn** [comment]
> Am I missing something? Supporting v2 does help support ARM? :relaxed:

Since compose v2 is in GO, there's native arm support. I know there was arm support from the pip package version of docker-compose v1, but it wasn't officially supported by docker, hence not having an arm release in the official packages.

It's just all-around better in my experience so far.

**Comment by spantaleev** [comment]
#4285 is still a problem, for which I've [found some workaround](https://github.com/mailcow/mailcow-dockerized/issues/4285#issuecomment-1031553935).

I don't use IPv6, so I'm not hitting any other issues besides #4285 with Compose v2 + `compose-switch` so far.

----

Another minor issue (annoyance) with doing `docker-compose up` with Compose v2 is what I've fixed in https://github.com/mailcow/mailcow-dockerized/pull/4457

**Comment by aminvakil** [comment]
How about setting both `HTTP_BIND` and `HTTPS_BIND` by default to `0.0.0.0`?

**Comment by MAGICCC** [comment]
Something like that then
```diff
- "${HTTPS_BIND:-:}:${HTTPS_PORT:-443}:${HTTPS_PORT:-443}"
- "${HTTP_BIND:-:}:${HTTP_PORT:-80}:${HTTP_PORT:-80}"
+ "${HTTPS_BIND:-0.0.0.0}:${HTTPS_PORT:-443}:${HTTPS_PORT:-443}"
+ "${HTTP_BIND:-0.0.0.0}:${HTTP_PORT:-80}:${HTTP_PORT:-80}"
```

Would need to check if something will break obv.

**Comment by aminvakil** [comment]
> Something like that then
> 
> ```diff
> - "${HTTPS_BIND:-:}:${HTTPS_PORT:-443}:${HTTPS_PORT:-443}"
> - "${HTTP_BIND:-:}:${HTTP_PORT:-80}:${HTTP_PORT:-80}"
> + "${HTTPS_BIND:-0.0.0.0}:${HTTPS_PORT:-443}:${HTTPS_PORT:-443}"
> + "${HTTP_BIND:-0.0.0.0}:${HTTP_PORT:-80}:${HTTP_PORT:-80}"
> ```
> 
> Would need to check if something will break obv.

It didn't break for me and worked just fine.

**Comment by aminvakil** [comment]
I'm not sure about these though:
https://github.com/mailcow/mailcow-dockerized/blob/880a68d563fb03aa4f80e0046448ab9ffab80ffe/update.sh#L701-L703
which has been added in
https://github.com/mailcow/mailcow-dockerized/commit/f3f6fb890857f49a25dc675a9d327e33e8183b21 and I cannot realize what has been broken.

**Comment by aminvakil** [comment]
Ah, I think something will break if `0.0.0.0` is used for IPv6... I'm not sure though.

**Comment by aminvakil** [comment]
Seems like there is a problem:
```
mailcowdockerized-nginx-mailcow-1       "/docker-entrypoint.…"   nginx-mailcow       running             0.0.0.0:80->80/tcp, 0.0.0.0:443->443/tcp
```
while
```
mailcowdockerized-postfix-mailcow-1     "/docker-entrypoint.…"   postfix-mailcow     running             0.0.0.0:25->25/tcp, 0.0.0.0:465->465/tcp, 0.0.0.0:587->587/tcp, :::25->25/tcp, :::465->465/tcp, :::587->587/tcp
```
(note the missing `:::80>80/tcp` where postfix has `:::25>25/tcp`)
but I've tested and it works (I can curl its IPv6, so :shrug:).

**Comment by philipp-tty** [comment]
Is there any update on v2 support? Considering v2.5.0 is bundled in the current docker ce/desktop install. Having to install v1 in parallel is a bit awkward.

**Comment by andryyy** [comment]
I hope nobody saw you doing this. Must have been really awkward. :)

**Comment by aminvakil** [comment]
@Cerberus-ik I'm using docker-compose v2 and has no one problem with it expect IPv6 binding which has been fixed using this PR (https://github.com/mailcow/mailcow-dockerized/pull/4531) as well.

I have tested and verified what I've done in the PR, but I don't know how test it using helper scripts as mentioned by @andryyy has said in the PR.

**Comment by philipp-tty** [comment]
> @Cerberus-ik I'm using docker-compose v2 and has no one problem with it expect IPv6 binding which has been fixed using this PR (#4531) as well.
> 
> I have tested and verified what I've done in the PR, but I don't know how test it using helper scripts as mentioned by @andryyy has said in the PR.

So I either have to make the same changes as you or wait for it to get merged into the main branch?

**Comment by philipp-tty** [comment]
> I hope nobody saw you doing this. Must have been really awkward. :)

Having to install deprecated eol "legacy" software for one application is awkward.

**Comment by hugalafutro** [comment]
> Having to install deprecated eol "legacy" software for one application is awkward.

I concur, out of my 10+ physical/virtual machines, the one running mailcow is the only one where I have to maintain 2 versions of docker-compose (even if it's well documented and easy with compose-switch).

**Comment by philipp-tty** [comment]
I think this is especially a high priority problem because v1 will only get 4 more months of critical bug fixes and severe security updates. So every sane sysadmin will remove v1 before the end of the year.

**Comment by gsloop** [comment]
> I hope nobody saw you doing this. Must have been really awkward. :)

Every time I watch the nominal "leader" of this project "snark" on something they think stupid, I am reminded once again, why I've waited over the last year - wanting to adopt MC as our mail server for a fairly large organization, but see, over and over again, what sure appears as very unprofessional/unhelpful behavior.

I'd have signed up for (and paid for) support, but have simply held off, continuing to hope that I'd get better feelings about the project - that I could have confidence in committing a lot of time and effort to moving the current mail server over.

Yet each time I think I'm "past" those nagging feelings of "I'll feel stupid if I commit all the time and effort to using MC for a really important, critical and highly visible project, simply to watch it all suck me into a professional disaster..." I find those feelings/intuition coming up again. (and again, and again.)

There are some good people at work on MC, and there are way too many who go a long way to discrediting the whole project. (I'm not singling out you, Andre - but you are often not helpful, IMO.) 

And why drop this here? Because that "comment" is just another example of what makes me incredibly wary, and I suspect, others too.

**Comment by andryyy** [comment]
Too bad. :)
